### PR TITLE
[FBFW-112] Territory Highlighting

### PR DESCRIPTION
**Pull request by shelbycodazen**
This PR adds filters to enable highlighting territories based on country control.

### Task
https://codazen.atlassian.net/browse/FBFW-112

### Testing Link
** The demo controls haven been removed, the filters can be applied manually by editing the [following attributes](https://github.com/codazen/webDiplomacy/pull/63/files/38d81fcc7d82e635ac54f54a0f295f6a4e9dcd29#diff-bb071e4c8244b7bf1638b4c55c2b0caf11df357208ac2f2aafc6f5b266778ea0R21-R23) on the path within the territory SVG.

```filter="url(#Russia)" fill-opacity="0.4" stroke-opacity="0.5"```

https://incredible-gaufre-d62647.netlify.app/beta/

### Demo Video
https://user-images.githubusercontent.com/99832938/160692283-d90ea92e-9032-46b0-833c-ac843a66803a.mov

### Screenshot
Screenshot

### CODE REVIEW

File: beta-src/src/data/map/land/LandTerritoriesMapData.ts [hunk 1]
```diff
@@ -35,6 +35,7 @@ export const ANKARA: TerritoryMapData = new TerritoryMapDataGenerator({
     x: 60,
     y: 100,
   },
+  fill: Texture.TURKEY,
```
Comment (bryancodazen) [inline beta-src/src/data/map/land/LandTerritoriesMapData.ts:38]: We don't want to set the initial fill to a map highlight, you can use the `texture` property to specific initial texture. Take a look at the interface.

File: beta-src/src/enums/Texture.ts [hunk 2]
```diff
@@ -1,7 +1,14 @@
 enum Texture {
+  AUSTRIA = "url(#austria)",
+  ENGLAND = "url(#england)",
+  FRANCE = "url(#france)",
+  GERMANY = "url(#germany)",
+  ITALY = "url(#italy)",
   LAND = "#F9F2E6",
-  SEA = "#96B2B2",
```
Comment (bryancodazen) [inline beta-src/src/enums/Texture.ts:3]: Was this an accident?

**Review by bryancodazen — CHANGES_REQUESTED** [review]
The map looks drastically different from the current map build in https://test.webdiplomacy.net, not sure if this is an accident or intentional to show the highlights but it makes it difficult to know what the highlight would look like on the current map. 

The highlights look pretty good! However, a couple of things: 
1. Is there any way to apply a radial gradient into the color, right now it looks like just a basic fill. This is not very noticeable on the smaller territories but very noticeable on the larger ones.
2. There is a whitish border that is mostly visible on the Russian color, it is mostly prominent when you zoom in - any way to remove this?
3. I noticed the filter code is identical to the one from the envision, what was the updates that needed to be made to make the highlight work? 

Overall it feels like this is really close, but we need to see the filter applied to the existing map. It's be cool too if you could setup a test function similar to the one here https://fbfw-61-map-componetization.netlify.app/ to see how the various colors would be applied. 

Screen Shot 2022-03-28 at 11 14 16 AM

File: beta-src/src/data/map/land/LandTerritoriesMapData.ts (hunk 1 above)
Comment (shelbycodazen) [reply beta-src/src/data/map/land/LandTerritoriesMapData.ts:38]: Would the following be okay or should areas that be be captured not have a fill defined? 

```
fill: Texture.LAND,
...
texture: {
 texture: Texture.TURKEY
}
```

File: beta-src/src/enums/Texture.ts (hunk 2 above)
Comment (shelbycodazen) [reply beta-src/src/enums/Texture.ts:3]: a-z'd these, it's still there.

File: beta-src/src/components/map/components/WDTerritory.tsx [hunk 3]
```diff
@@ -34,8 +35,9 @@ const WDTerritory: React.FC<WDTerritoryProps> = function ({
       )}
       <path
         d={territoryMapData.path}
-        fill={territoryMapData.fill}
-        fillOpacity={territoryMapData.texture?.texture ? 0.2 : 1}
+        fill={Texture.LAND}
+        fillOpacity={territoryMapData.texture?.texture ? 0.5 : 1}
+        filter={territoryMapData.fill}
```
[the remark this answers is not in the document]
Comment (shelbycodazen) [reply beta-src/src/components/map/components/WDTerritory.tsx:40]: Oh man, the naming here doesn't help that.  This is what enabled the highlighting to work, but also broke the coloring on the rest of the map.

```fill={Texture.LAND}```  - the filter needs a fill defined in order to work.

```filter={territoryMapData.fill}``` - ```territoryMapData.fill``` here is the actual filter def, not the fill above.


I've changed this to the following:
```
        fill={territoryMapData.fill}
        fillOpacity={territoryMapData.texture?.texture ? 0.5 : 1}
        filter={territoryMapData.texture?.texture}
```

This works and gets the rest of the map to the right color, so long as this is also okay to do: https://github.com/codazen/webDiplomacy/pull/63#discussion_r836878464

File: beta-src/src/components/map/WDMap.tsx [hunk 4]
```diff
@@ -44,6 +44,160 @@ const WDMap: React.ForwardRefExoticComponent<
...
+            in="SourceGraphic"
+            in2="offset-blur"
+            result="inverse"
+          />
+          <feFlood floodColor="#FC4343" floodOpacity="1" result="color" />
+          <feComposite operator="in" in="color" in2="inverse" result="shadow" />
+          <feComponentTransfer in="shadow" result="shadow">
+            <feFuncA type="linear" slope="2.5" />
+          </feComponentTransfer>
+        </filter>
```
[the remark this answers is not in the document]
Comment (shelbycodazen) [reply beta-src/src/components/map/WDMap.tsx:68]: It feels like that should be doable.  The id of the filter changes too, for each country.

File: beta-src/src/data/map/land/LandTerritoriesMapData.ts (hunk 1 above)
Comment (bryancodazen) [reply beta-src/src/data/map/land/LandTerritoriesMapData.ts]: Anything defined in this file is for the initial map load. No territory is captured initially until we get the API which shows which countries own what territories.

File: beta-src/src/components/map/components/WDTerritory.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (bryancodazen) [reply beta-src/src/components/map/components/WDTerritory.tsx]: What path is this for? The control or texture path? 

I think you need to define these normally in the initial map data and then create a function to apply this highlighting and show an example of its use.

[the remark this answers is not in the document]
Comment (shelbycodazen) [reply beta-src/src/components/map/components/WDTerritory.tsx]: This is for the control path.  I was able to remove the `filter` prop and apply it with a function.

File: beta-src/src/components/map/WDMap.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (shelbycodazen) [reply beta-src/src/components/map/WDMap.tsx:68]: Got something working for this.  Just needs some refactoring with TypeScript.

**Comment by shelbycodazen** [comment]
> The map looks drastically different from the current map build in https://test.webdiplomacy.net, not sure if this is an accident or intentional to show the highlights but it makes it difficult to know what the highlight would look like on the current map.

Resolved!

> 1. Is there any way to apply a radial gradient into the color, right now it looks like just a basic fill. This is not very noticeable on the smaller territories but very noticeable on the larger ones.

Still trying to solve that part. I was able to adjust the filters, but then it looks terrible on the small territories. 

> 2. There is a whitish border that is mostly visible on the Russian color, it is mostly prominent when you zoom in - any way to remove this?

I adjusted the floodOpactity on the `russia` filter only, it looks like it goes away.

> 3. I noticed the filter code is identical to the one from the envision, what was the updates that needed to be made to make the highlight work?

Have a utility function that does this now.  I just need to get it working with TypeScript.

> Overall it feels like this is really close, but we need to see the filter applied to the existing map. It's be cool too if you could setup a test function similar to the one here https://fbfw-61-map-componetization.netlify.app/ to see how the various colors would be applied.

The demo link is updated to allow passing in a COUNTRY and TERRITORY name to apply the filter.

**Review by nebevets — APPROVED** [review]
LGTM

File: beta-src/src/utils/map/WDCountryHighlightFilters.tsx [hunk 5]
```diff
@@ -0,0 +1,72 @@
+import * as React from "react";
+import webDiplomacyTheme from "../../webDiplomacyTheme";
+
+interface CountryHighlightFilter {
+  countryName: string;
```
Comment (bryancodazen) [inline beta-src/src/utils/map/WDCountryHighlightFilters.tsx:5]: We should make this a country enum, and then reverse map the country enum when setting the id.

File: beta-src/src/utils/map/WDCountryHighlightFilters.tsx [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  floodColor: string;
```
Comment (bryancodazen) [inline beta-src/src/utils/map/WDCountryHighlightFilters.tsx:6]: I don't think we should pass this, we should reference the country color based on the web dip theme using the country enum.

File: beta-src/src/utils/map/applyCountryHighlight.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
+/**
+ * This function shows which attributes need to be udpated/applied to set the highlighting properly.  They should be set on the WDTerritory element when it is time to wire this up with the global state.
+ * @param countryName
+ * @param territoryName
+ * @returns
+ */
+
+export default function applyCountryHighlight(
+  countryName: string,
+  territoryName: string,
```
Comment (bryancodazen) [inline beta-src/src/utils/map/applyCountryHighlight.ts:10]: Both of these should be based on their respective enums.

File: beta-src/src/utils/map/applyCountryHighlight.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+  );
+
+  if (territoryElement == null) {
+    alert("Territory not found!"); // eslint-disable-line
+    return;
+  }
+
+  territoryElement.setAttribute("filter", `url(#${countryName.toLowerCase()})`);
+  territoryElement.setAttribute("fill-opacity", "0.4");
+  territoryElement.setAttribute("stroke-opacity", "0.5");
```
Comment (bryancodazen) [inline beta-src/src/utils/map/applyCountryHighlight.ts:23]: Let's remove the alert and only apply the highlight if the element is valid. We can incorporate error messaging into the UI bottom left status panel once that is implemented.

**Review by bryancodazen — CHANGES_REQUESTED** [review]
AWESOME WORK on the highlights! Just a couple of changes.

**Review by bryancodazen — APPROVED** [review]
LGTM

**Pull request merged**